core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbRecordIterator.java

```diff
@@ -126,73 +126,83 @@ class LmdbRecordIterator implements RecordIterator {
 
 	@Override
 	public long[] next() {
-		long stamp = txnLock.readLock();
-		try {
-			if (closed) {
-				log.debug("Calling next() on an LmdbRecordIterator that is already closed, returning null");
-				return null;
+		long stamp = txnLock.tryOptimisticRead();
+		long[] value = nextInternal();
+		if (!txnLock.validate(stamp)) {
+			stamp = txnLock.readLock();
+			try {
+				// Retry read
+				value = nextInternal();
```

Comment:
Don't we risk that calling nextInternal() twice will skip an element if the write happened right after the first call to nextInternal() and right before validating the lock?

Reply:
Agh, yeah, you are right, that will be an issue. Let me have a look at what I can do about this.

Reply:
I think I was too trigger-happy with an optimistic lock here. We should try something else here (e.g., the `ReadPrefReadWriteLockManager` you proposed), but IMO this is for a separate PR.